src/merkle_tree/constraints.rs

```diff
@@ -101,75 +45,250 @@ where
         let ns = cs.into();
         let cs = ns.cs();
         f().and_then(|val| {
-            let mut path = Vec::new();
-            for &(ref l, ref r) in val.borrow().path.iter() {
-                let l_hash = HGadget::OutputVar::new_variable(
-                    ark_relations::ns!(cs, "l_child"),
-                    || Ok(l),
+            let leaf_sibling = LeafH::OutputVar::new_variable(
+                ark_relations::ns!(cs, "leaf_sibling"),
+                || Ok(val.borrow().leaf_sibling_hash.clone()),
+                mode,
+            )?;
+            let leaf_position_bit = Boolean::new_variable(
+                ark_relations::ns!(cs, "leaf_position_bit"),
+                || Ok(val.borrow().leaf_index & 1 == 1),
+                mode,
+            )?;
+            let pos_list = val.borrow().position_list();
+            let mut path = Vec::with_capacity(pos_list.len() - 1);
+            for &bit in &pos_list[..pos_list.len() - 1] {
+                // we do not include leaf bit here, so ignore last element returned by pos_list
+                path.push(Boolean::new_variable(
+                    ark_relations::ns!(cs, "path_bit"),
+                    || Ok(bit),
                     mode,
-                )?;
-                let r_hash = HGadget::OutputVar::new_variable(
-                    ark_relations::ns!(cs, "r_child"),
-                    || Ok(r),
+                )?)
+            }
+
+            let mut auth_path = Vec::with_capacity(val.borrow().auth_path.len());
+            for v in val.borrow().auth_path.iter() {
+                auth_path.push(TwoToOneH::OutputVar::new_variable(
+                    ark_relations::ns!(cs, "auth_path_node"),
+                    || Ok(v.clone()),
                     mode,
-                )?;
-                path.push((l_hash, r_hash));
+                )?)
             }
-            Ok(PathVar { path })
+            Ok(PathVar {
+                path,
+                auth_path,
+                leaf_sibling,
+                leaf_position_bit,
+            })
         })
     }
 }
 
+impl<P, LeafH, TwoToOneH, ConstraintF> PathVar<P, LeafH, TwoToOneH, ConstraintF>
+where
+    P: Config,
+    LeafH: CRHGadget<P::LeafHash, ConstraintF>,
+    TwoToOneH: TwoToOneCRHGadget<P::TwoToOneHash, ConstraintF>,
+    ConstraintF: Field,
+{
+    /// Given a leaf, calculate the root as if the merkle has this leaf at this path.
+    fn calculate_root(
+        &self,
+        leaf_hash_parameter: &LeafH::ParametersVar,
+        two_to_one_hash_parameter: &TwoToOneH::ParametersVar,
```

Comment:
```suggestion
        leaf_hash_params: &LeafH::ParametersVar,
        two_to_one_hash_params: &TwoToOneH::ParametersVar,
```

Reply:
fixed. I've also renamed all `*parameter` to `params` to consistency